x-pack/plugins/canvas/server/lib/essql_strategy.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { from } from 'rxjs';
+import { map, zipObject } from 'lodash';
+
+import { ISearchStrategy, PluginStart } from 'src/plugins/data/server';
+
+import { EssqlSearchStrategyRequest, EssqlSearchStrategyResponse } from '../../types';
+
+import { buildBoolArray } from '../../common/lib/request/build_bool_array';
+import { sanitizeName } from '../../common/lib/request/sanitize_name';
+import { normalizeType } from '../../common/lib/request/normalize_type';
+interface CursorResponse {
+  cursor?: string;
+  rows: string[][];
+}
+
+type QueryResponse = CursorResponse & {
+  columns: Array<{
+    name: string;
+    type: string;
+  }>;
+};
+
+export const essqlSearchStrategyProvider = (
+  data: PluginStart
+): ISearchStrategy<EssqlSearchStrategyRequest, EssqlSearchStrategyResponse> => {
+  return {
+    search: (request, options, { esClient }) => {
+      const { count, query, filter, timezone, params } = request;
+
+      const searchUntilEnd = async () => {
+        let response = await esClient.asCurrentUser.transport.request({
+          path: '/_sql?format=json',
+          method: 'POST',
+          body: {
+            query,
+            params,
+            time_zone: timezone,
+            fetch_size: count,
+            client_id: 'canvas',
+            filter: {
+              bool: {
+                must: [{ match_all: {} }, ...buildBoolArray(filter)],
+              },
+            },
+          },
+        });
+        let body = response.body as QueryResponse;
```

Comment:
I think we can get rid of this cast by doing something like `asCurrentUser.sql.query<QueryResponse>` or something similar.